manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,445 @@
+import logging
+from abc import ABC, abstractmethod
+from eth_typing import ChecksumAddress, URI
+from io import TextIOBase
+from typing import Dict, Optional, Set, Union
+from urllib.parse import ParseResult, urlparse
+from web3 import Web3
+from ..core.smtlib import Array, BitVec, BitVecConstant, BitVecITE, BitVecZeroExtend, ConstraintSet
+from ..ethereum.state import State
+from ..exceptions import EthereumError
+
+logger = logging.getLogger(__name__)
+
+
+# sam.moelius: map records which (symbolic) offsets have been written to.  data holds the values
+# written.
+
+
+class Storage:
+    def __init__(self, constraints: ConstraintSet, address: int):
+        self.constraints = constraints
+        self.warned = False
+        self.map = constraints.new_array(
+            index_bits=256,
+            value_bits=1,
+            name=f"STORAGE_MAP_{address:x}",
+            avoid_collisions=True,
+            default=0,
+        )
+        self.data = constraints.new_array(
+            index_bits=256,
+            value_bits=256,
+            name=f"STORAGE_DATA_{address:x}",
+            avoid_collisions=True,
+            default=0,
+        )
+        self.dirty = False
+
+    def set(self, offset: Union[int, BitVec], value: Union[int, BitVec]):
+        self.map[offset] = 1
```

Comment:
Maybe  `value != 0` instead of  `1` ?

Reply:
I think it should be 1.  If the underlay is a `RemoteWorldState` and the value at that offset is non-zero, then the `map` should reflect that the overlay differs from the underlay at that offset.

Reply:
So...
`self.data.is_known(offset)`  is already doing map's work then?
And `len(self.data.written)` the work of self.dirty ?

Reply:
> So...
> `self.data.is_known(offset)` is already doing map's work then?

Sort of. I think `self.data.is_known(offset)` should hold iff `self.map.is_known(offset)`. So, in that sense, map and data's `written` fields do the same work.

The advantage to splitting them is that it allows to construct this conditional:
https://github.com/trailofbits/manticore/blob/1d26902093080cbf236e3b6db8ed351eae09cab7/manticore/platforms/evm_world_state.py#L54
It might be possible to do away with map and use `self.data.written` or `self.data.is_known` directly. I am trying to respect the encapsulation properties of `self.data`.

> And `len(self.data.written)` the work of self.dirty ?

I think this may be true.

Reply:
https://github.com/trailofbits/manticore/blob/1d26902093080cbf236e3b6db8ed351eae09cab7/manticore/platforms/evm_world_state.py#L54
`BitVecITE(256, self.map[offset] != 0, self.data[offset], default) ` !=?  `self.data.get(offset, default) `

Reply:
Indeed, it looks like you are right!  I eliminated the use of `self.map`.